disputer/index.js

```diff
@@ -34,7 +34,8 @@ async function run(price, address, shouldPoll, pollingDelay) {
 
   // Client and dispute bot
   const empClient = new ExpiringMultiPartyClient(Logger, ExpiringMultiParty.abi, web3, emp.address);
-  const gasEstimator = new GasEstimator(Logger);
+  const getTime = () => empClient.getLastUpdateTime();
```

Comment:
This change could make sense, but there's some subtle dependencies that this creates between the clients:
1. If the empClient only allows itself to update every 60 seconds, but a client that uses getTime can update every 10 seconds, the second client will end up only updating every 60 seconds because the time will be the same until it jumps by 60+ seconds. We don't have update throttling logic in the empClient right now, so this wouldn't be an issue unless we added it.
2. If another client is updated *before* the empClient, its time will effectively be based on the last iteration of the empClient. I think this means we would probably want to call update on empClient *first* in every loop.

Would 1 or 2 be particularly bad for any client? Also, are there any clients that are super sensitive to how we compute time?

There's no problem with this dependence, but to keep the time computation working as expected is a bit subtle -- we should definitely document these caveats IMO. There are also other options like just forcing all clients to *compute* the time the same way (system time vs contract time) or querying the time outside of these clients and using getTime to just return that externally computed time.

Not super opinionated about how we do this -- just wanted to give some random thoughts.

Reply:
I think having an external module solely responsible for keeping track of time would be the best solution. It'd make testing and debugging much easier and also easier to reason about. 

Re: system time vs contract time, I'm very much in favor of using contract time as that's the time system that the contracts use. WDYT ? @chrismaree

Reply:
Actually, its probably much easier to just make `this.getTime` an asynchronous method for each client and we can pass in `empContract.getCurrentTime()` instead of `empClient.getLastUpdateTime()`

Reply:
done

Reply:
So I think that the `GasEstimator` is one exception to the time rule. In all other situations, the time the implementor uses should be the last time the client updated. For the `GasEstimator` it should _always_ be the current time. This is because the `GasEstimator` is used when sending transactions on main net and as such we want to know the current gas price _right now_, not what it was the last time the client updated. 

We do have rate-limiting in the `GasEstimator` such that at most the value is 60 seconds old. So this means that we _dont_ want to use the time source from the `empClient.getLastUpdateTime()`. 

With that said, I'm not convinced we want to call `empContract.getCurrentTime()` for this. This means an extra blockchain call to infura every single time we run the gasEstimator. What difference is there between just using the local machine time for this? 

Specifically, this module is used to infer information about the external state of the chain and what an appropriate price is. As a result, I see no difference between using `empContract.getCurrentTime()` & `Math.floor(Date.now() / 1000)` other than the former executes an extra async call every time.

My preference, therefore, is to use `Math.floor(Date.now() / 1000)` in the Gas estimator and use `empClient.getLastUpdateTime()` in all other contexts.

Reply:
I just don't like to have two different sources of time because it makes reasoning about tests more difficult. However, it is true that GasEstimator and the EMP do not interact. Therefore I'm OK with using Date.now(), since gasPrice is determined outside of the blockchain state and is set by human market participants

Reply:
using `Date.now()` would also be consistent with the `getTime` that is fed into `priceFeed`, right @mrice32 ?